### PR TITLE
Update known_hosts module to better handle @cert-authority keys

### LINKED ISSUE #70339: known_hosts module breaks when using @cert-authority keys
##### SUMMARY

If I attempt to add a key for `host.example.com` ansible removes my `@cert-authority *.example.com`

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
known_hosts

##### ANSIBLE VERSION

```paste below
ansible-playbook 2.9.10
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.7.3 (default, Dec 20 2019, 18:57:59) [GCC 8.3.0]
```

##### CONFIGURATION

Completely default

##### OS / ENVIRONMENT

Docker image on amd64 running debian:buster-slim    image id: 43e3995ee54a
Installed via pip

##### STEPS TO REPRODUCE

play.yml

    - hosts: localhost
      gather_facts: no
      vars:
        known_hosts_path:
        example_com_ed25519_key: >
          host.example.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFp8VtD59XAcxkj1qbfCtB1in9nm5WiipORjtVJUBA6I
      tasks:

      - name: Create a copy of known_hosts
        copy:
          src: existing_known_hosts
          dest: test_known_hosts

      - name: add the ed25519 host key
        known_hosts:
          name: host.example.com
          key: "{{ example_com_ed25519_key }}"
          state: present
          path: "test_known_hosts"
        register: result

      - name: get the file content
        command: "cat test_known_hosts"
        register: known_hosts_v1

      - name: assert that the key was added and ordering preserved
        assert:
          that:
          - 'result is changed'
          - 'known_hosts_v1.stdout_lines[-1].strip() == example_com_ed25519_key.strip()'
          - 'known_hosts_v1.stdout_lines[4].startswith("@cert-authority")'

existing_known_hosts

    example.com ssh-dss AAAAB3NzaC1kc3MAAACBALT8YHxZ59d8yX4oQNPbpdK9AMPRQGKFY9X13S2fp4UMPijiB3ETxU1bAyVTjTbsoag065naFt13aIVl+u0MDPfMuYgVJFEorAZkDlBixvT25zpKyQhI4CtHhZ9Y9YWug4xLqSaFUYEPO31Bie7k8xRfDwsHtzTRPp/0zRURwARHAAAAFQDLx2DZMm3cR8cZtbq4zdSvkXLh0wAAAIAalkQYziu2b5dDRQMiFpDLpPdbymyVhDMmRKnXwAB1+dhGyJLGvfe0xO+ibqGXMp1aZ1iC3a/vHTpYKDVqKIIpFg5r0fxAcAZkJR0aRC8RDxW/IclbIliETD71osIT8I47OFc7vAVCWP8JbV3ZYzR+i98WUkmZ4/ZUzsDl2gi7WAAAAIAsdTGwAo4Fs784TdP2tIHCqxAIz2k4tWmZyeRmXkH5K/P1o9XSh3RNxvFKK7BY6dQK+h9jLunMBs0SCzhMoTcXaJq331kmLJltjq5peo0PnLGnQz5pas0PD7p7gb+soklmHoVp7J2oMC/U4N1Rxr6g9sv8Rpsf1PTPDT3sEbze6A== root@freezer
    @cert-authority *.example.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIDXh1gk2xgS2MekPvo7ZEKiOT7HoyvOAzai2GqoLXGHO
    # example.net ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIM6OSqweGdPdQ/metQaf738AdN3P+itYp1AypOTgXkyj root@localhost

Output

    # ansible-playbook play.yml --diff
    [WARNING]: No inventory was parsed, only implicit localhost is available
    [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

    PLAY [localhost] *************************************************************************************************************
    ...
    TASK [add the ed25519 host key] **********************************************************************************************
    --- before: test_known_hosts
    +++ after: test_known_hosts
    @@ -1,3 +1,3 @@
     example.com ssh-dss AAAAB3NzaC1kc3MAAACBALT8YHxZ59d8yX4oQNPbpdK9AMPRQGKFY9X13S2fp4UMPijiB3ETxU1bAyVTjTbsoag065naFt13aIVl+u0MDPfMuYgVJFEorAZkDlBixvT25zpKyQhI4CtHhZ9Y9YWug4xLqSaFUYEPO31Bie7k8xRfDwsHtzTRPp/0zRURwARHAAAAFQDLx2DZMm3cR8cZtbq4zdSvkXLh0wAAAIAalkQYziu2b5dDRQMiFpDLpPdbymyVhDMmRKnXwAB1+dhGyJLGvfe0xO+ibqGXMp1aZ1iC3a/vHTpYKDVqKIIpFg5r0fxAcAZkJR0aRC8RDxW/IclbIliETD71osIT8I47OFc7vAVCWP8JbV3ZYzR+i98WUkmZ4/ZUzsDl2gi7WAAAAIAsdTGwAo4Fs784TdP2tIHCqxAIz2k4tWmZyeRmXkH5K/P1o9XSh3RNxvFKK7BY6dQK+h9jLunMBs0SCzhMoTcXaJq331kmLJltjq5peo0PnLGnQz5pas0PD7p7gb+soklmHoVp7J2oMC/U4N1Rxr6g9sv8Rpsf1PTPDT3sEbze6A== root@freezer
    -@cert-authority *.example.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIDXh1gk2xgS2MekPvo7ZEKiOT7HoyvOAzai2GqoLXGHO
     # example.net ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIM6OSqweGdPdQ/metQaf738AdN3P+itYp1AypOTgXkyj root@localhost
    +host.example.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFp8VtD59XAcxkj1qbfCtB1in9nm5WiipORjtVJUBA6I

    changed: [localhost]

    TASK [get the file content] **************************************************************************************************
    changed: [localhost]

    TASK [assert that the key was added and ordering preserved] ******************************************************************
    fatal: [localhost]: FAILED! => {"msg": "The conditional check 'known_hosts_v1.stdout_lines[4].startswith(\"@cert-authority\")' failed. The error was: error while evaluating conditional (known_hosts_v1.stdout_lines[4].startswith(\"@cert-authority\")): list object has no element 4"}

    PLAY RECAP *******************************************************************************************************************
    localhost                  : ok=3    changed=3    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

##### EXPECTED RESULTS

I expected adding `hosts.example.com` not to have the side effect of removing `@cert-authority *.example.com ...`

##### ACTUAL RESULTS

See Above

### PR DESCRIPTION
**Pull request by zoredache**
##### SUMMARY

Fixes #70339 
An attempt at fixing the known_hosts bug I submitted, plus updating tests to better describe the problem.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
known_hosts

### CODE REVIEW

**Review by Akasurde — CHANGES_REQUESTED** [review]
Can you please add a changelog entry for this change [here](https://github.com/ansible/ansible/tree/devel/changelogs/fragments/)?

File: lib/ansible/modules/known_hosts.py [hunk 1]
```diff
@@ -247,12 +247,22 @@ def search_for_host_key(module, host, key, path, sshkeygen):
...
             found_key = normalize_known_hosts_key(l)
-            if new_key['host'][:3] == '|1|' and found_key['host'][:3] == '|1|':  # do not change host hash if already hashed
-                new_key['host'] = found_key['host']
-            if new_key == found_key:  # found a match
-                return True, False, found_line  # found exactly the same key, don't replace
-            elif new_key['type'] == found_key['type']:  # found a different key for the same key type
-                return True, True, found_line
+
+            if 'options' in found_key and found_key['options'][:15] == '@cert-authority':
+                if found_key['host'] != new_key['host']:
```
Comment (Akasurde) [inline lib/ansible/modules/known_hosts.py]: Can we remove this condition?

Comment (zoredache) [reply lib/ansible/modules/known_hosts.py]: You are right those extra bits were not required. They were left over from me trying to find where the problem was.

**Comment by zoredache** [comment]
Updated to remove the extraneous bits Akasurde noticed. Also added some tests for `@revoked` signed host keys.

**Comment by Akasurde** [comment]
@zoredache Can you please take a look into this failure - https://app.shippable.com/github/ansible/ansible/runs/172199/38/console ? Thanks.

**Comment by zoredache** [comment]
@Akasurde, I could use some assistance with out to handle that. I am not certain, but I suspect that is CentOS instance is running the default version of OpenSSH for Centos6, which was 5.3 (release 2009-10-01).  The functionality to support the certificates was added in 5.4 (release 2010-03-08).

Is there some obvious way to just not run the tests on Centos6 that hasn't been updated to a newer version of OpenSSH? I assume there must be some way to not run the tests on systems where a module wouldn't work?

**Comment by zoredache** [comment]
> @zoredache This PR was evaluated as a potentially problematic PR for the following reasons:

Yup, sorry, I was trying to rebase my branch, but merged accidentally, I have correctly reset and re pushed with only my changes.

File: changelogs/fragments/known_hosts_cert-authority_keys.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - known_hosts - Fix issue with `@cert-authority` entries in known_hosts incorrectly being removed.
```
Comment (webknjaz) [inline changelogs/fragments/known_hosts_cert-authority_keys.yml:2]:
```suggestion
  - known_hosts - Fix issue with ``@cert-authority`` entries in ``known_hosts`` incorrectly being removed.
```

**Comment by bcoca** [comment]
@zoredache any chance you can fix the conflict and pass the tests?
!waiting_on_contributor

**Review by Akasurde — APPROVED** [review]

**Pull request merged**